python/predict_urls/prediction_funcs.py

```diff
@@ -0,0 +1,280 @@
+                            row[k] = v
+                # Append to OPD_Source_table.csv
+                df = pd.read_csv(OPD_SOURCE_TABLE)
+                df = pd.concat([df, pd.DataFrame([row])], ignore_index=True)
+                df.to_csv(OPD_SOURCE_TABLE, index=False)
+                if verbose:
+                    print(f"Added to OPD_Source_table: {new_url}")
+                continue
+            else:
+                if verbose:
+                    print(f"{new_url}: not valid. Skipping.")
+    return None
+
+def auto_update_sources(
+    outdated_days=None,
+    verbose=False
+):
+    """
+    Automatically check for new data by incrementing year in URLs for testable sources. Adds valid URLs to OPD_Source_table.csv.
+
+    Args:
+        outdated_days (int): How many days before a source is considered outdated.
+        verbose (bool): Print progress.
+    """
+    # 1. Load table and parse last_coverage_check as datetime
+    df = pd.read_csv(OPD_SOURCE_TABLE)
+    df["last_coverage_check_dt"] = pd.to_datetime(df["last_coverage_check"], errors="coerce")
+    # 2. Group by composite key and get max last checked date
+    composite_key = ['State', 'SourceName', 'AgencyFull', 'TableType']
+    max_df = df.loc[df.groupby(composite_key)["Year"].idxmax()]
+    # 3. Filter URLs with exactly one 4-digit year
+    max_df = max_df[max_df["URL"].str.contains(r"\d{4}", regex=True)]
+    max_df = max_df[max_df["URL"].str.count(r"\d{4}") == 1]
+    # Keep rows where last_coverage_check_dt is NaT or older than outdated_days
+    current_date = datetime.now()
+    if outdated_days is not None:
+        cutoff_date = current_date - pd.Timedelta(days=outdated_days)
+        to_test = max_df[max_df["last_coverage_check_dt"] <= cutoff_date]
+    else:
+        last_year = current_date.year - 1
+        to_test = max_df[max_df["last_coverage_check_dt"].dt.year <= last_year]
+    to_test = to_test.drop(columns=["last_coverage_check_dt"])
+    # 6. For each outdated row, try to find new URLs by incrementing year using try_url_years
+    current_year = datetime.now().year
+    count = 0
+    for row in to_test.itertuples(index=False):
+        url = row.URL
+        data_type = row.DataType.lower()
+        year_match = re.search(r"\d{4}", url)
+        if not year_match:
+            continue
+        year_str = year_match.group()
+        year = int(year_str)
+        spreadsheet_fields = row._asdict()
+        if year < current_year:
+            for y in range(year + 1, current_year + 1):
+                valid, new_url = find_valid_url_for_year(url, y, year_str, data_type, spreadsheet_fields)
+                if valid:
+                    new_row = spreadsheet_fields.copy()
+                    new_row["URL"] = new_url
+                    new_row["Year"] = str(y)
+                    new_row["last_coverage_check"] = datetime.now().strftime("%m/%d/%Y")
+                    new_row["coverage_start"] = f"01/01/{y}"
```

Comment:
Other columns whose values might change year-to-year:
1. *source_url*: This will change in most cases. It's easy enough to find the source_url manually so that might be better than writing something that tries to find it. Maybe we should just leave it empty, which will cause an error in testing if it's not filled in.
2. *readme*: I'm comfortable assuming the readme is unchanged
3. *date_field*: If populated, we could read in some data to check if the date_fields is still a valid column.
4. *agency_field*: See date_field